### PR TITLE
Reshape op

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
It adds new feature requested by the community:
 * Reshape operator with ability to reinterpret input tensor(s) without copying

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
    * Added the operator which uses ShareData to pass same memory to the output, while reinterpreting the shape and/or layout
 - What was changed, added, removed?
    * Added the operator and tests
    * Removed excessive checks for buffer size
 - What is most important part that reviewers should focus on?
    * reshape.cc
 - Was this PR tested? How?
    * python test
 - Were docs and examples updated, if necessary?
    * added auto-generated docs for the operator

**JIRA TASK**: [DALI-XXXX]

### CODE REVIEW

**Comment by mzient** [comment]
!build

File: dali/pipeline/operators/util/reshape.cc [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+      const kernels::TensorListView<kernels::StorageCPU, Integer> &shape) {
+  DALI_ENFORCE(shape.sample_dim() == 1 || (shape.sample_dim() == 2 && shape.num_samples() == 1),
+    "Reshape: shape input must be a list of 1D tensors or a single 2D tensor");
+  if (shape.sample_dim() == 2) {
+    auto shape_tensor = shape[0];
+    int N = shape_tensor.shape[0];
+    int dim = shape_tensor.shape[1];
+    output_shape_.resize(N, dim);
+    for (int i = 0; i < N; i++) {
+      for (int d = 0; d < dim; d++) {
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.cc]: could cache the tensor_shape_span outside of the second loop

File: dali/pipeline/operators/util/reshape.cc [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+        output_shape_.tensor_shape_span(i)[d] = *shape_tensor(i, d);
+      }
+    }
+  } else {
+    int N = shape.num_samples();
+    int sample_dim;
+    for (int i = 0; i < N; i++) {
+      int current_sample_dim = shape.tensor_shape_span(i)[0];
+      if (i) {
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.cc]: maybe a more explicit `i > 0`

File: dali/pipeline/operators/util/reshape.cc [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        output_shape_.tensor_shape_span(i)[d] = shape.tensor_data(i)[d];
+      }
+    }
+  }
+}
+
+template <typename Backend>
+template <typename TensorListLike>
+void Reshape<Backend>::ShapeFromInput(const TensorListLike &tl) {
+  TYPE_SWITCH(tl.type().id(), type2id, type, (int32_t),
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.cc]: don't you want to support other integral types? If not, please rephrase the DALI_FAIL below to say that int32 is expected

File: dali/pipeline/operators/util/reshape.cc [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+  DALI_ENFORCE(output_shape_.num_samples() == N,
+    "Reshape: the new shape mush have same number of samples. Got " +
+    to_string(output_shape_.num_samples()) + ", expected " + to_string(N));
+
+  for (int i = 0; i < N; i++) {
+    auto actual_volume = volume(input_shape_.tensor_shape_span(i));
+    auto requested_volume = volume(output_shape_.tensor_shape_span(i));
+    DALI_ENFORCE(actual_volume == requested_volume,
+      "Input and output samples should have the same number of elements");
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.cc]: you could also show `actual_volume` vs `requested_volume` and the sample id

File: dali/pipeline/operators/util/reshape.cc [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+  }
+}
+
+template <typename Backend>
+TensorLayout Reshape<Backend>::GetOutputLayout(const Workspace &ws) const {
+  if (!layout_.empty()) {
+    DALI_ENFORCE(output_shape_.sample_dim() == layout_.ndim(), "Reshape: requested layout '" +
+      layout_.str() + "' not compatible with " + to_string(output_shape_.sample_dim()) + "D shape");
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.cc]: what does `D` mean here?

File: dali/pipeline/operators/util/reshape.cc [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+  auto in_layout = this->InputLayout(ws, 0);
+  return in_layout.ndim() == output_shape_.sample_dim() ? in_layout : TensorLayout();
+}
+
+template <>
+void Reshape<CPUBackend>::RunImpl(HostWorkspace &ws) {
+  auto &out = ws.OutputRef<CPUBackend>(0);
+  auto &in = ws.InputRef<CPUBackend>(0);
+  out.Resize(output_shape_);
+  out.SetContiguous(false);
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.cc]: out of curiosity: why do you need to set this? Could it be contiguous if the input was ?

File: dali/pipeline/operators/util/reshape.h [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+template <typename Backend>
+class Reshape : public Operator<Backend> {
+ public:
+  using Base = Operator<Backend>;
+  using Workspace = workspace_t<Backend>;
+
+  explicit Reshape(const OpSpec &spec_);
+
+  bool CanInferOutputs() const override { return false; }
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.h]: didn't you mean
```suggestion
  bool CanInferOutputs() const override { return true; }
```
here?

File: dali/pipeline/operators/util/reshape.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+  bool SetupImpl(std::vector<OutputDesc> &output_desc, const Workspace &ws) override;
+
+  void RunImpl(Workspace &ws) override;
+
+ private:
+  kernels::TensorListShape<> input_shape_, output_shape_;
+  kernels::TensorShape<> uniform_shape_;
+  TensorLayout layout_;
+  enum class ShapeSource {
```
Comment (jantonguirao) [inline dali/pipeline/operators/util/reshape.h]: I'd rather separate the enum definition from the member variable declaration

File: dali/test/python/test_operator_reshape.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        self.reshape = ops.Reshape(device = device, shape = (224, 224 * 3));
+
+    def define_graph(self):
+        jpegs, labels = self.input(name = "Reader")
+        images = self.resize(self.decode(jpegs))
+        if self.device == "gpu":
+          images = images.gpu()
+          reshaped = self.reshape(images)
+        else:
+          reshaped = self.reshape(images)
```
Comment (jantonguirao) [inline dali/test/python/test_operator_reshape.py]:
```
images = images.gpu() if self.device == 'gpu' else images
reshaped = self.reshape(images)
```

File: dali/test/python/test_operator_reshape.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+        return [images, reshaped]
+
+def verify(imgs, reshaped):
+  for i in range(len(imgs)):
+    assert imgs.at(i).shape == (224, 224, 3)
+    assert reshaped.at(i).shape == (224, 224 * 3)
+    assert_array_equal(imgs.at(i).flatten(), reshaped.at(i).flatten())
+
+
```
Comment (jantonguirao) [inline dali/test/python/test_operator_reshape.py]: To follow what we do in other tests (it looks nice when you run the tests with `nosetests --verbose`)
```
def check_reshape(device, batch_size):
  pipe = ReshapePipeline(device, batch_size)
  pipe.build()
  for iter in range(10):
    imgs, reshaped = pipe.run()
    if device == 'gpu'
      imgs = imgs.as_cpu()
      reshaped = reshaped.as_cpu()
    verify(imgs, reshaped)

def test_reshape():
  for device in ['cpu', 'gpu']:
    for batch_size in [16]:
      yield check_reshape, device, batch_size
```

File: dali/pipeline/operators/util/reshape.cc (hunk 5 above)
Comment (mzient) [reply dali/pipeline/operators/util/reshape.cc]: After building the string it is 1D, 2D, 3D, etc.

File: dali/pipeline/operators/util/reshape.cc (hunk 6 above)
Comment (mzient) [reply dali/pipeline/operators/util/reshape.cc]: Does the comment address it?

File: dali/pipeline/operators/util/reshape.cc [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+
+namespace dali {
+
+DALI_SCHEMA(Reshape)
+  .DocStr(R"code(Treats content of the input as if it had a different shape and layout.)code")
+  .NumInput(1, 2)
+  .NumOutput(1)
+  .AddOptionalArg<int>("shape", "The desired shape of the output. Number of elements in "
+                                "each sample must match that of the input sample.",
+                                std::vector<int>(), true)
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/pipeline/operators/util/reshape.cc]: That's future-proof - for more-than 1D tensor arguments.

**Comment by mzient** [comment]
!build

File: dali/python/backend_impl.cc [hunk 12]
```diff
@@ -70,13 +70,26 @@ static string TensorLayoutRepr(const TensorLayout &tl) {
...
 void ExposeTensorLayout(py::module &m) {
-  py::class_<TensorLayout>(m, "TensorLayout")
+  auto &tl = py::class_<TensorLayout>(m, "TensorLayout")
   .def(py::init([](string s) {
     return new TensorLayout(s);
   }))
   .def("__str__", &TensorLayout::str)
   .def("__repr__", TensorLayoutRepr)
   .def("__len__", &TensorLayout::ndim);
+  #define DEFINE_LAYOUT_CMP(name, expr)\
```
Comment (jantonguirao) [inline dali/python/backend_impl.cc]: move macros to the top of the file

Comment (mzient) [reply dali/python/backend_impl.cc]: It's not a good idea - this only makes sense in this function. What I can do (and will) is to add `#undef` after use.

**Comment by mzient** [comment]
!build

**Pull request merged**